lib/ocpp/v201/smart_charging.cpp

```diff
@@ -77,36 +77,58 @@ ProfileValidationResultEnum SmartChargingHandler::validate_tx_profile(const Char
  * - K01.FR.20
  * - K01.FR.34
  * - K01.FR.43
- * - K01.FR.45
  * - K01.FR.48
  */
-ProfileValidationResultEnum SmartChargingHandler::validate_profile_schedules(const ChargingProfile& profile) const {
-    auto schedules = profile.chargingSchedule;
+ProfileValidationResultEnum
+SmartChargingHandler::validate_profile_schedules(ChargingProfile& profile,
+                                                 std::optional<EvseInterface*> evse_opt) const {
+    auto schedules = &profile.chargingSchedule;
 
-    for (auto schedule : schedules) {
+    for (ChargingSchedule& schedule : *schedules) {
         // A schedule must have at least one chargingSchedulePeriod
         if (schedule.chargingSchedulePeriod.empty()) {
             return ProfileValidationResultEnum::ChargingProfileNoChargingSchedulePeriods;
         }
 
-        auto charging_schedule_period = schedule.chargingSchedulePeriod[0];
-
         for (auto i = 0; i < schedule.chargingSchedulePeriod.size(); i++) {
+            auto charging_schedule_period = &schedule.chargingSchedulePeriod[i];
             // K01.FR.19
-            if (charging_schedule_period.numberPhases != 1 && charging_schedule_period.phaseToUse.has_value()) {
+            if (charging_schedule_period->numberPhases != 1 && charging_schedule_period->phaseToUse.has_value()) {
                 return ProfileValidationResultEnum::ChargingSchedulePeriodInvalidPhaseToUse;
             }
 
             // K01.FR.31
-            if (i == 0 && charging_schedule_period.startPeriod != 0) {
+            if (i == 0 && charging_schedule_period->startPeriod != 0) {
                 return ProfileValidationResultEnum::ChargingProfileFirstStartScheduleIsNotZero;
                 // K01.FR.35
-            } else if (i != 0) {
-                auto next_charging_schedule_period = schedule.chargingSchedulePeriod[i];
-                if (next_charging_schedule_period.startPeriod <= charging_schedule_period.startPeriod) {
+            } else if (i != 0 && i + 1 < schedule.chargingSchedulePeriod.size()) {
```

Comment:
Shouldn't we check this for `i == 0` too since you check the next period against the current